independent-projects/arc/runtime/src/main/java/io/quarkus/arc/impl/ArcContainerImpl.java

```diff
@@ -158,6 +162,17 @@ public ArcContainerImpl(CurrentContextFactory currentContextFactory, boolean str
         instance = InstanceImpl.forGlobalEntrypoint(Object.class, Collections.emptySet());
 
         this.beans = List.copyOf(beans);
+        this.beansByRawType = Map.copyOf(beansByRawType);
```

Comment:
A simple heap analysis shows that using the immutable map from  JDK (`Map.copyOf()`) reduces the retained size of this map from 15KB to 9KB for a quarkus application that has ~ 250 beans.

Reply:
please don't, the jdk Map.of is less efficient for lookups

Reply:
The lookup performance is more than good enough for us in this particular case.

Reply:
this map isn't used at runtime? Otherwise really it has a serious scalability limitation as it's not aligning the array sizes to multiples of two - indeed that saves a little memory but it requires using modulo operators on each lookup - that will completely stall a core if hyperthreading is used, which is very common

Reply:
this map isn't used at runtime? Otherwise really it has a serious scalability limitation as it's not aligning the array sizes to multiples of two - indeed that saves a little memory but it requires using modulo operators on each lookup - that will completely stall a core if hyperthreading is used, which is very common

Reply:
It is used at runtime but (a) it's not used in a hot path and (b) we measured the performance with a microbenchmark and the results were more or less the same :shrug:.

Reply:
Ok well remember:
 - we had to patch Arc in the past to remove some of these as they had a signifcant impact on benchmarks
 - you won't notice on micro benchmarks unless you're using all CPUs
So I'd personally avoid them but you know better how much of an "hot path" this is..

Reply:
> we had to patch Arc in the past to remove some of these as they had a signifcant impact on benchmarks

I do remember quite a few optimizations I'm not exactly proud of :shrug:.

> you won't notice on micro benchmarks unless you're using all CPUs
So I'd personally avoid them

All benchmarks are tricky because they run in a specific system configuration. But I strongly disagree that we should avoid microbenchmarks . In an ideal world, we would start with a microbenchmark and continue with benchmarking several quarkus app configurations (small app, big app, etc.)..

> you know better how much of an "hot path" this is..

I'm pretty sure here.

Reply:
>All benchmarks are tricky because they run in a specific system configuration. But I strongly disagree that we should avoid microbenchmarks 

sorry I didn't mean "avoid microbenchmarks" - not at all, that's great :)
I meant we should avoid Map.of for anything that's going to be frequently read, as while I agree there's a memory benefit it's a tradeoff which is strongly skewed to having an extreme cost on CPU for just a very small amount of memory.

> I'm pretty sure here.

All good then, thanks :)